src/test/chainstate_write_tests.cpp

```diff
@@ -42,4 +42,58 @@ BOOST_FIXTURE_TEST_CASE(chainstate_write_interval, TestingSetup)
     BOOST_CHECK(sub->m_did_flush);
 }
 
+// Test that we do PERIODIC flushes inside ActivateBestChain.
+// This is necessary for reindex-chainstate to be able to periodically flush
+// before reaching chain tip.
+BOOST_FIXTURE_TEST_CASE(write_during_multiblock_activation, TestChain100Setup)
+{
+    struct TestSubscriber final : CValidationInterface {
+        const CBlockIndex* m_tip{nullptr};
+        const CBlockIndex* m_flushed_at_block{nullptr};
+        void ChainStateFlushed(ChainstateRole, const CBlockLocator&) override
+        {
+            m_flushed_at_block = m_tip;
+        }
+        void UpdatedBlockTip(const CBlockIndex* block_index, const CBlockIndex*, bool) override {
+            m_tip = block_index;
+        }
+    };
+
+    auto& chainstate{Assert(m_node.chainman)->ActiveChainstate()};
+    BlockValidationState state_dummy{};
+
+    // Pop two blocks from the tip
+    const CBlockIndex* tip{chainstate.m_chain.Tip()};
+    CBlockIndex* second_from_tip{tip->pprev};
+
+    {
+        LOCK(m_node.chainman->GetMutex());
+        LOCK(chainstate.MempoolMutex());
+        chainstate.DisconnectTip(state_dummy, nullptr);
+        chainstate.DisconnectTip(state_dummy, nullptr);
+    }
+
+    BOOST_CHECK(second_from_tip->pprev == chainstate.m_chain.Tip());
+
+    // Set m_next_write to current time
+    SetMockTime(GetTime());
+    chainstate.FlushStateToDisk(state_dummy, FlushStateMode::ALWAYS);
+    m_node.validation_signals->SyncWithValidationInterfaceQueue();
+    // The periodic flush interval is between 50 and 70 minutes (inclusive)
+    // The next call to a PERIODIC write will flush
+    SetMockTime(GetTime<std::chrono::minutes>() + 70min);
+
+    const auto sub{std::make_shared<TestSubscriber>()};
+    m_node.validation_signals->RegisterSharedValidationInterface(sub);
+
+    // ActivateBestChain back to tip
+    chainstate.ActivateBestChain(state_dummy, nullptr);
+    BOOST_CHECK(tip == chainstate.m_chain.Tip());
+    // Check that we flushed inside ActivateBestChain while we were at the
+    // second block from tip, since FlushStateToDisk is called with PERIODIC
+    // inside the outer loop.
+    m_node.validation_signals->SyncWithValidationInterfaceQueue();
+    BOOST_CHECK(sub->m_flushed_at_block == second_from_tip);
```

Comment:
👍 I can confirm that this passes after the PR and fails without the PERIODIC flushing moved inside the loop